service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureExtractor.java

```diff
@@ -0,0 +1,89 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import static uk.nhs.adaptors.gp2gp.ehr.utils.CodeableConceptMappingUtils.extractTextOrCoding;
+import static uk.nhs.adaptors.gp2gp.ehr.utils.DateFormatUtil.toHl7Format;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hl7.fhir.dstu3.model.AllergyIntolerance;
+import org.hl7.fhir.dstu3.model.DateTimeType;
+import org.hl7.fhir.dstu3.model.Extension;
+import org.hl7.fhir.dstu3.model.StringType;
+
+import uk.nhs.adaptors.gp2gp.ehr.utils.CodeableConceptMappingUtils;
+import uk.nhs.adaptors.gp2gp.ehr.utils.DateFormatUtil;
+
+public class AllergyStructureExtractor {
+    private static final String ALLERGY_REASON_ENDED = "Reason Ended: ";
+    private static final String ALLERGY_REASON_END_URL = "reasonEnded";
+    private static final String ALLERGY_END_DATE_URL = "endDate";
+    private static final String ALLERGY_NO_INFO = "No information available";
+    private static final String REACTION = "Reaction %S";
+    private static final String REACTION_DESCRIPTION = " Description: ";
+    private static final String REACTION_SEVERITY = " Severity: ";
+    private static final String REACTION_EXPOSURE_ROUTE = " Exposure Route: ";
+    private static final String REACTION_MANIFESTATIONS = " Manifestation(s): ";
+    private static final String COMMA = ", ";
+
+    public static String extractReasonEnd(Extension extension) {
+        var reasonEnd = extension.getExtension()
+            .stream()
+            .filter(value -> value.getUrl().equals(ALLERGY_REASON_END_URL))
+            .findFirst()
+            .map(Extension::getValue)
+            .map(reason -> ((StringType) reason).getValueAsString())
+            .orElse(StringUtils.EMPTY);
+
+        if (!reasonEnd.isEmpty() && !reasonEnd.equals(ALLERGY_NO_INFO)) {
+            return ALLERGY_REASON_ENDED + reasonEnd;
+        }
+
+        return StringUtils.EMPTY;
+    }
+
+    public static String extractEndDate(Extension extension) {
+        return extension.getExtension()
+            .stream()
+            .filter(allergyEnd -> allergyEnd.getUrl().equals(ALLERGY_END_DATE_URL))
+            .findFirst()
+            .map(value -> (DateTimeType) value.getValue())
+            .map(DateFormatUtil::toHl7Format)
+            .orElse(StringUtils.EMPTY);
+    }
+
+    public static String extractOnsetDate(AllergyIntolerance allergyIntolerance) {
+        if (allergyIntolerance.hasOnset() && allergyIntolerance.getOnsetDateTimeType().hasValue()) {
+            return toHl7Format(allergyIntolerance.getOnsetDateTimeType());
+        }
+        return StringUtils.EMPTY;
+    }
+
+    public static String extractReaction(AllergyIntolerance.AllergyIntoleranceReactionComponent reactionComponent,
```

Comment:
What do you think about using builder pattern in complicated reaction/note building logic?
https://refactoring.guru/design-patterns/builder

Something like that:

```
 public static String extractReaction(AllergyIntolerance.AllergyIntoleranceReactionComponent reactionComponent,
        AtomicInteger reactionCount) {
        return new ReactionBuilder(reactionCount.getAndIncrement())
            .description(reactionComponent.getDescription()) // returns null if empty
            .severity(reactionComponent.getSeverity()) // retruns null if empty
            .exposureRoute(reactionComponent.getExposureRoute()) // created new empty CodeableConcept if empty
            .manifestation(reactionComponent.getManifestation()) // created new empty list if empty
            .build();
    }

    public static class ReactionBuilder {
        private static final String REACTION = "Reaction %S";
        private static final String REACTION_DESCRIPTION = " Description: ";
        private static final String REACTION_SEVERITY = " Severity: ";
        private static final String REACTION_EXPOSURE_ROUTE = " Exposure Route: ";
        private static final String REACTION_MANIFESTATIONS = " Manifestation(s): ";
        private static final String COMMA = ", ";

        private String reaction;

        public ReactionBuilder(int increment) {
            reaction = String.format(REACTION, increment);
        }

        public ReactionBuilder description(String description) {
            reaction += Optional.ofNullable(description)
                .map(value -> REACTION_DESCRIPTION + value)
                .orElse(StringUtils.EMPTY);

            return this;
        }

        public ReactionBuilder severity(AllergyIntolerance.AllergyIntoleranceSeverity severity) {
            reaction += Optional.ofNullable(severity)
                .map(value -> REACTION_SEVERITY + value)
                .orElse(StringUtils.EMPTY);

            return this;
        }

        public ReactionBuilder exposureRoute(CodeableConcept exposureRoute) {
            reaction += CodeableConceptMappingUtils.extractTextOrCoding(exposureRoute)
                .map(value -> REACTION_EXPOSURE_ROUTE + value)
                .orElse(StringUtils.EMPTY);
                
            return this;
        }

        public ReactionBuilder manifestation(List<CodeableConcept> manifestation) {
            String joinedManifestation = manifestation
                .stream()
                .map(CodeableConceptMappingUtils::extractTextOrCoding)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.joining(COMMA));

            if (StringUtils.isNoneBlank(joinedManifestation)) {
                reaction += REACTION_MANIFESTATIONS + joinedManifestation;
            }

            return this;
        }

        public String build() {
            return reaction;
        }
    }
```

Reply:
or similar way to retrievePertinentInformation in AllergyStructureMapper?